Source/WebCore/rendering/CounterNode.cpp

```diff
@@ -139,11 +140,15 @@ CounterNode* CounterNode::previousInPreOrder() const
 
 int CounterNode::computeCountInParent() const
 {
+    auto sumOrFirstValue = [](int firstValue, int secondValue) {
+        return WTF::signedAddInt32Overflows(firstValue, secondValue, secondValue) ? firstValue : secondValue;
+    };
+
```

Comment:
@darinadler @cdumez Do we have some common pattern in the codebase for this? Blink seems to use a checkedAdd helper for instance.

Reply:
We probably should not use `signedAddInt32Overflows`. That’s an implementation detail of `SaturatedArithmetic.h`, not designed for use outside the header. Ideally that header would be changed to just use `sumOverflows` from `CheckedArithmetic.h` to avoid duplications. But to preserve the performance `sumOverflows` would likely need to be specialized or optimized so it’s super efficient for two `int32_t` values. We’d want to do that, though. We don’t want two functions that do the same thing, one with a nicer name but a poorer implementation.

This all is a saturated arithmetic header, and `saturatedSum` will clamp to the maximum `int32_t` value quite efficiently. But perhaps the counter specification requires we not saturate but just not add a new value at all if it would overflow. Too bad because saturated arithmetic is a nice solution for such things.

To improve the aesthetic of checked arithmetic it would be fine to evolve the checked arithmetic header to have the kind of function we need. We have `checkedSum` and `sumOverflows` and the `Checked` class template. Interestingly, `checkedSum` is _almost_ what we want, in fact it even stores the first value if adding the second value overflows. But you can’t get that value out without the code intentionally crashing in the `Checked::value` function!

I think that having `sumOrFirstValue` as a local lambda here is not great since this seems like the kind of operation we’d want to use more than once. I’d want it to be in `CheckedArithmetic.h`. We didn’t design that header, and it seems it’s falling short here. We may be able to come up with an even better name for that operation too. It’s possible that the `checkedAdd` function is something Chrome folks added to `CheckedArithmetic.h` after they forked WebKit for this very reason, but I don’t think that "add" vs. "sum" is very clear.

Reply:
I've moved it to `CheckedArithmetic.h` and renamed it to `sumIfNoOverflow()`. I'm not sure about this name as it doesn't clearly tells what is returned if the sum won't take place. I'm open to suggestions!